File: src/pramp/ShiftedArraySearch.java
package pramp;

import java.util.Arrays;

public class ShiftedArraySearch {
    static int shiftedArrSearch2(int[] shiftArr, int num) {
        for (int i = 0; i < shiftArr.length; i++) {
            if (shiftArr[i] == num) return i;
        }

        return -1;
    }

    static int shiftedArrSearch(int[] shiftArr, int num) {
        // 1. find the offset point:
        int p = 0;
        int q = shiftArr.length;
        int m = 0;

        while (p <= q) {
            m = p + (q - p) / 2;
            int mid = shiftArr[m];

            if (m > 0 && mid < shiftArr[m - 1]) break; // m is the offset point
            if (mid < shiftArr[p]) p = m + 1;
            else q = m - 1;
        }

        // 2. Find the correct portion to do bin search
        if (num >= shiftArr[m] && num <= shiftArr[shiftArr.length - 1]) {
            p = m;
            q = shiftArr.length;
        } else {
            p = 0;
            q = m;
        }

        // 3. Do the bin search on the correct portion and return the index
        int index = Arrays.binarySearch(shiftArr, p, q, num);
        return index >= 0 ? index : -1;
    }

    public static void main(String[] args) {
        System.out.println(shiftedArrSearch(new int[]{2}, 2));
        System.out.println(shiftedArrSearch(new int[]{1, 2}, 2));
        System.out.println(shiftedArrSearch(new int[]{0, 1, 2, 3, 4, 5}, 1));
        // TODO fix this one
        System.out.println(shiftedArrSearch(new int[]{1, 2, 3, 4, 5, 0}, 0));
        System.out.println(shiftedArrSearch(new int[]{9, 12, 17, 2, 4, 5}, 17));
        System.out.println(shiftedArrSearch(new int[]{9, 12, 17, 2, 4, 5, 6}, 4));
    }
}

File: src/main/java/pramp/ShiftedArraySearch.kt
package main.java.pramp

fun shiftedArrSearch2(shiftArr: IntArray, num: Int): Int {
    for (i in shiftArr.indices) {
        if (shiftArr[i] == num) return i
    }

    return -1
}

fun shiftedArrSearch(shiftArr: IntArray, num: Int): Int {
    var p = 0
    var q = shiftArr.size - 1
    var m = 0
    while (p <= q) {
        m = p + (q - p) / 2
        if (m > 0 && shiftArr[m] < shiftArr[m - 1]) break // pivot point is found
        if (shiftArr[m] > shiftArr[q]) p = m + 1
        else q = m - 1
    }

    if (num >= shiftArr[m] && num <= shiftArr.last()) {
        // right side
        p = m
        q = shiftArr.size - 1
    } else {
        // left side
        p = 0
        q = m - 1
    }

    val result = shiftArr.binarySearch(num, p, q + 1)
    return if (result >= 0) result else -1
}

fun main() {
//    println(shiftedArrSearch2(intArrayOf(2), 2))
//    println(shiftedArrSearch(intArrayOf(2), 2))
//    println(shiftedArrSearch2(intArrayOf(1, 2), 2))
//    println(shiftedArrSearch(intArrayOf(1, 2), 2))
//    println(shiftedArrSearch2(intArrayOf(0, 1, 2, 3, 4, 5), 1))
//    println(shiftedArrSearch(intArrayOf(0, 1, 2, 3, 4, 5), 1))

    // TODO fix this one
//    println(shiftedArrSearch2(intArrayOf(1, 2, 3, 4, 5, 0), 0))
    println(shiftedArrSearch(intArrayOf(1, 2, 3, 4, 5, 0), 0))
    println(shiftedArrSearch2(intArrayOf(9, 12, 17, 2, 4, 5), 17))
    println(shiftedArrSearch(intArrayOf(9, 12, 17, 2, 4, 5), 17))
    println(shiftedArrSearch2(intArrayOf(9, 12, 17, 2, 4, 5, 6), 4))
    println(shiftedArrSearch(intArrayOf(9, 12, 17, 2, 4, 5, 6), 4))
}